deepakarorawins/CucumberFramework
# Sharing Test Context between Cucumber Step Definitions

How to Sharing Test Context
between Cucumber Step
Definitions using PicoContainer
We will be performing below steps to share data state across steps:
1. Add PicoContainer to the Project
2. Create a Test Context class which will hold all the objects state
3. Divide the Steps class into multiple steps classes with logical
separation
4. Write Constructor to share Test Context
Step 1: Add PicoContainer Library to the
Maven Project
This is really simple, as we have been using Maven Project, all we need to do is to add the dependencies into the project POM file. Dependencies information can be taken from Maven Repository – Cucumber PicoContainer.

Note: Make sure to add these dependencies under <dependencies>Add here</dependencies> tag. Also, it also suggested using the same version as a cucumber. In my case, it is 1.2.5.
Step 2: Create a Test Context class
Be wise to create this class logically. Just think what all the information your Steps file is using and put that information into this class. In our case our steps file is just using the below information:
• PageObjects : Provided by PageObjectManager
• WebDriver: Provided by WebDriverManager
• Properties: Provided by FileReaderManager
So, the case is simple. We just need the above objects in our Test Context class. But if you re-look at the objects, you would release that our FileReaderManager is already a Singleton Class and to use it we don’t need to create an instance of it. It creates its instance by itself. So no need to add FileReaderManager to TestContext class, as this class can be referred directly statically like FileReaderManager.getInstance().
1. Create a New Package and name it as cucumber, by right click on the src/test/java and select New >> Package. We will keep all the Cucumber Helper classes in the same package moving forward.
2. Create a New Class file and name it as TestContext by right click on the above created package and select New >> Class. TestContext.java

Step 3: Divide the Steps file
I would just divide the steps file as I did the separations between the Page Objects. For every different page, we have a
separate PageObject class. So it makes sense to have a separate step definition class for every page as well. May be your application is different and you won’t get to agree with my approach. In that case who cares 🙂 Divide your steps accordingly.
216
 
Create four New Classes in the stepDefinitions package with the
following names:
• HomePageSteps
• ProductPageSteps
• CartPageSteps
• CheckoutPageSteps
I am sure you won’t need my help to do this separation. Just start copying pasting information from steps class into the above-created classes accordingly. You would get the code in the below step.
Step 4: Write Constructor for Step Definition
classes to share TestContext
Let’s start with the HomePageSteps. The current state of the HomePageSteps class is without Constructor:

I hope now you would understand, what I was referring to. You
need WebDriverManager & PageObjectManager in every step file, otherwise, you need to create objects for both classes using a new operator again and again, which is Kill Bill.
Now with just adding Constructor to HomePageSteps file and
pass TestContext as a Parameter to constructor would take all the pain. Within the TestContext object, we have everything available which is required for the test. So now let’s see how the new HomePageSteps class would look.

Isn’t this looks simple and lovely? Did you realize how much code we just reduced from the class?
Using PicoContainer to share state between steps in a scenario is easy and non-intrusive. All you need is a constructor that requires an object that PicoContainer can create and inject. PicoContainer is invisible. Add a dependency to cucumber-picocontainer and make sure that the constructors for the step classes require an instance of the same class.
Now you can start copying the code for the rest of the step classes.

Note: Delete the Steps class from stepDefinitions package, as it is no longer required.

Run the Cucumber Test
Run as JUnit
Now we are all set to run the Cucumber test. Right,
Click on TestRunner class and Click Run As >> JUnit
Test. Cucumber will run the script the same way it runs in Selenium WebDriver and the result will be shown in the left-hand side project explorer window in JUnit tab.
<img width="403" alt="Screenshot 2020-03-17 at 6 27 18 PM" src="https://user-images.githubusercontent.com/30565176/76916198-00181380-687d-11ea-9bfe-fb32ea5a9f8c.png">
